flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DefaultResourceAllocationStrategy.java

```diff
@@ -110,9 +121,93 @@ public ResourceAllocationResult tryFulfillRequirements(
+            TaskManagerResourceInfoProvider taskManagerResourceInfoProvider) {
+        ResourceProfile requiredRedundantResources =
+                totalResourceProfile.multiply(redundantTaskManagerNum);
+        ResourceReleaseResult.Builder builder = ResourceReleaseResult.builder();
+
+        List<TaskManagerInfo> taskManagersIdleTimeout = new ArrayList<>();
+        List<TaskManagerInfo> taskManagersNonTimeout = new ArrayList<>();
+        long currentTime = System.currentTimeMillis();
+        taskManagerResourceInfoProvider
+                .getRegisteredTaskManagers()
+                .forEach(
+                        taskManagerInfo -> {
+                            if (taskManagerInfo.isIdle()
+                                    && currentTime - taskManagerInfo.getIdleSince()
+                                            >= taskManagerTimeout.toMilliseconds()) {
+                                taskManagersIdleTimeout.add(taskManagerInfo);
+                            } else {
+                                taskManagersNonTimeout.add(taskManagerInfo);
+                            }
+                        });
+
+        List<PendingTaskManager> pendingTaskManagersNonUse = new ArrayList<>();
+        List<PendingTaskManager> pendingTaskManagersInuse = new ArrayList<>();
+        taskManagerResourceInfoProvider
+                .getPendingTaskManagers()
+                .forEach(
+                        pendingTaskManager -> {
+                            if (pendingTaskManager.getPendingSlotAllocationRecords().isEmpty()) {
+                                pendingTaskManagersNonUse.add(pendingTaskManager);
+                            } else {
+                                pendingTaskManagersInuse.add(pendingTaskManager);
+                            }
+                        });
+
+        if (taskManagersIdleTimeout.isEmpty() && pendingTaskManagersNonUse.isEmpty()) {
+            // short-cut for nothing to release
+            return builder.build();
+        }
+
+        ResourceProfile resourcesToKeep = ResourceProfile.ZERO;
+
+        // check whether available resources of used (pending) task manager is enough.
+        ResourceProfile availableResourcesOfNonIdle =
+                getAvailableResourceOfTaskManagers(taskManagersNonTimeout);
+        resourcesToKeep = resourcesToKeep.merge(availableResourcesOfNonIdle);
+        if (!canFulfillRequirement(requiredRedundantResources, resourcesToKeep)) {
+            ResourceProfile availableResourcesOfNonIdlePendingTaskManager =
+                    getAvailableResourceOfPendingTaskManagers(pendingTaskManagersInuse);
+            resourcesToKeep = resourcesToKeep.merge(availableResourcesOfNonIdlePendingTaskManager);
+        }
+
+        // try reserve or release unused (pending) task managers
+        for (TaskManagerInfo taskManagerInfo : taskManagersIdleTimeout) {
+            if (canFulfillRequirement(requiredRedundantResources, resourcesToKeep)) {
```

Comment:
Minor: we can use a local variable to skip unnecessary `canFulfillRequirement` calls. `if (redundantFulfilled || canFulfillRequirement(xxx, xxx))`.